crates/wasmtime/src/runtime/vm/continuation.rs

```diff
@@ -13,6 +13,33 @@ cfg_if::cfg_if! {
 /// once. The linearity is checked dynamically in the generated code
 /// by comparing the revision witness embedded in the pointer to the
 /// actual revision counter on the continuation reference.
+///
+/// In the optimized implementation, the continuation logically
+/// represented by a VMContObj not only encompasses the pointed-to
+/// VMContRef, but also all of its parents:
+///
+/// ```text
+///
+///                     +----------------+
+///                 +-->|   VMContRef    |
+///                 |   +----------------+
+///                 |            ^
+///                 |            | parent
+///                 |            |
+///                 |   +----------------+
+///                 |   |   VMContRef    |
+///                 |   +----------------+
+///                 |            ^
+///                 |            | parent
+///  last ancestor  |            |
```

Comment:
OK, so reading this drawing I may understand what you are on about with the ancestor business. What you call `last_ancestor` is really the initial frame of the delimited continuation that is being captured. I think we should name all this differently. The frame is not the "last" ancestor, as it may be adopted by another ancestor etc. However, it is the tail of the linked list that defines the continuation. Thus, I think it is better to use names such as "tail"/"head" to describe these objects. The data structure you are building is known as a "circular list". I think it is worthwhile to adopt this terminology to make the implementation easier to understand and more readily accessible to other people.

Reply:
I'm not particularly married to the name `last_ancestor`. However, this field is only ever set on the "head" continuation (bottom-most continuation in the picture) and only if it is suspended. Thus, if its set, it indeed points to the last ancestor.

I'm not particularly keen on the term `tail`, since it is somewhat overloaded in the context of linked lists. It is both used to refer to the next node in a linked list, but also to refer to the end of the list.

Reply:
I am happy to consider anything but last ancestor. Am I correct that the continuation object points to the bottom stack in the chain (i.e. the stack that executed the suspension)? If so, then I think we can optimise it ever-so-slightly but returning the "last ancestor" instead and set its parent to be the bottom stack. We should save a field that way too.

Reply:
That's possible, but I would prefer holding off from that. The reason is that it complicates how at `suspend`, we propagate the required information to the `resume` site.
In the current version of this PR, the following happens:
1. We identify the currently running continuation  `c1` by reading the stack chain field of the `VMContext`. This will be the bottom of the continuation chain of the continuation object we are about to construct.
2. The handler search gives us the other end of the chain, `c2`. We set `c1.last_ancestor` to `c2`.
3. We perform the stack switch.
4. Back in `resume`, we read `c1` from the `VMContext` (it's still saved in the stack chain field) and obtain the other end (`c2`) from its `last_ancestor` field.

In the design you propose, we would instead make `c2` point to `c1` in step 2. But this means that in the 4th step above, we have no way to obtain `c2`. We would have to either temporarily write `c2` into the `VMContext` or extend the `switch` instruction so that we can pass more payloads. I'd very much prefer avoiding that, at least for now.

Reply:
Ugh, it strikes again. I understand. We really need to get rid of the auxiliary structures. We should be able to encode everything in one chain. Certainly we need this before/for the upstreaming.

Reply:
The issue I described is orthogonal from the details of how we represent the chain of currently active continuations. In any design, there will always be a field in the `VMContext` that points to the currently active continuation. That's the field accessed in step 1 and 4 above.
In that sense, the design in this PR works without requiring any new channels (in the widest sense) to get information from the `suspend` to the `resume` site.

Reply:
I am not convinced. To me the current design/implementation has a certain spaghetti feeling to it. But it can be refactored later.